RealEstateDataCleaning.py

```diff
@@ -8,113 +8,160 @@
-	df = make_YearBuilt_bins(df, size=10)
-	df['Year'] = pd.DatetimeIndex(df['CloseDate']).year
-	df['Age'] = df['Year'] - df['YearBuilt']
-	df['AssociationFee'] = np.where((df['AssociationFee']> 0),1,0)
-	df['PoolonProperty'] = np.where((df['PoolonProperty'] !=  'None'),1,0)
-	df['Bedrooms'] = df['UpperBedrooms'] + df['MainBedrooms']
-	df['FullBaths'] = df['UpperFullBaths'] + df['MainFullBaths']
-	df['HalfBaths'] = df['UpperHalfBaths'] + df['MainHalfBaths']
-	df.drop(['StreetNumber', 'StreetName', 'City', 'CountyOrParish', 'StateOrProvince', 
-			 'PostalCode','UpperBedrooms', 'MainBedrooms', 'UpperFullBaths', 'MainFullBaths', 
-			 'UpperHalfBaths', 'MainHalfBaths'], axis=1, inplace=True)
-	df.to_csv(output_file)
+def add_upper_and_main(data_frame, desired):
+    """
+    Look for the "Upper" and "Main" values for the desired column.  Sum these together and
+    create a new column named desired.  Finally, remove the "Upper" and "Main" columns
+
+    e.g. "Bedrooms" is the sum of the "UpperBedrooms" and "MainBedrooms" values
+
+    :param data_frame: callback frame object to manipulate
+    :param desired: string denoting the desired domain
+    :return: None
+    """
+    upper = "Upper{}".format(desired)
+    main = "Main{}".format(desired)
+
+    data_frame[desired] = data_frame[upper] + data_frame[main]
+
+    data_frame.drop([upper, main], axis=1, inplace=True)
+
+def main(df):
+    # method does too much - change to be testable (see example)
+
+    # df is an unclear name.  I'm not familiar with pandas, and it took me awhile to figure out
+    # that it's a data frame
+
+    df.drop(df[df.SqFtTotal == 0].index, inplace=True)
+    df['CloseDate'] = pd.to_datetime(df['CloseDate'])
+    df = concatenate_address(df)
+
+    # next section gets better with map
+    df = extract_dummies(df, Exterior_features, col='Exterior')
+    df['BasementDesc'] = df['BasementDesc'].apply(Basement_cleanup)
+    df = extract_dummies(df, BasementDesc_features, col='BasementDesc')
+    df = extract_dummies(df, ParkingDesc_features, col='ParkingDesc')
+    df['Stories'] = df['Stories'].apply(Stories_cleanup)
+    df = extract_dummies(df, Stories_features, col='Stories')
+
+    # some whitespace will help in this method :)
+    df = make_YearBuilt_bins(df, size=10)
+    df['Year'] = pd.DatetimeIndex(df['CloseDate']).year
+    df['Age'] = df['Year'] - df['YearBuilt']
+    df['AssociationFee'] = np.where((df['AssociationFee']> 0),1,0)
+    df['PoolonProperty'] = np.where((df['PoolonProperty'] !=  'None'),1,0)
+
+    # whenever you see repeated operations, consider if this can be done in a function.
+    # if you read "add_upper_and_main" you'll see that it also does the drop function
+    # df['Bedrooms'] = df['UpperBedrooms'] + df['MainBedrooms']
+    # df['FullBaths'] = df['UpperFullBaths'] + df['MainFullBaths']
+    # df['HalfBaths'] = df['UpperHalfBaths'] + df['MainHalfBaths']
+    for domain in ['Bedrooms', 'FullBaths', 'HalfBaths']:
```

Comment:
You can see here how tight this looks by pushing the operations down into the function.  From the top level here we can quickly parse that we're adding together the "upper" and "main" values.  We can reduce the clutter here and ensure that the initial column gets dropped.

We also now have something is testable, which you can see in the next class.